### PR TITLE
Enable creditmemo feature for zero order total

### PR DESCRIPTION
**Pull request by viral-wagento**
### Description

When customer place an order of zero ordertotal using Zero Subtotal Checkout Payment method and Admin create an invoice - order goes into Processing. After we can't cancel order. 

At such scenario, requirement comes to create creditmemo and return ordered item back in stock.
Magento doesn't allow to create a creditmemo for zero order total. So creating credit memo is required to return ordered items back in stock.

### Fixed Issues (if relevant)

1. magento/magento2#13153 : Credit memo for orders with zero TotalPaid

### Manual testing scenarios

1. Place an order of Zero amount using Zero Subtotal Checkout Payment method.
2. Create an invoice for that order.
3. As per Magento behaviour, we can't see Credit Memo Option for Zero Total Order.
4. But Now, you can see 'Credit Memo' option, create credit memo for that order and return item back to stock.
5. We can check same steps for Order of amount greater than zero, It will work proper for them too.

### Actual Result
For Zero Order Total, we can't see Credit Memo Option Available, and so we are unable to generate it.
actual-result

### Expected Result
One should be able to generate Credit Memo even for Zero OrderTotal. 
expected-result

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/Sales/Controller/Adminhtml/Order/Creditmemo/Save.php [hunk 1]
```diff
@@ -33,21 +33,29 @@ class Save extends \Magento\Backend\App\Action
...
      * @param \Magento\Backend\Model\View\Result\ForwardFactory $resultForwardFactory
+     * @param \Psr\Log\LoggerInterface $logger
      */
     public function __construct(
         Action\Context $context,
         \Magento\Sales\Controller\Adminhtml\Order\CreditmemoLoader $creditmemoLoader,
         CreditmemoSender $creditmemoSender,
-        \Magento\Backend\Model\View\Result\ForwardFactory $resultForwardFactory
+        \Magento\Backend\Model\View\Result\ForwardFactory $resultForwardFactory,
+        \Psr\Log\LoggerInterface $logger
```
Comment (larsroettig) [inline app/code/Magento/Sales/Controller/Adminhtml/Order/Creditmemo/Save.php]: It is Major Pull Request pls change the taget branch or 

`\Psr\Log\LoggerInterface $logger = null`

https://devdocs.magento.com/guides/v2.2/extension-dev-guide/versioning/codebase-changes.html
New required constructor scalar argument (without pre-configured value)

**Review by larsroettig — CHANGES_REQUESTED** [review]
Hi @viral-wagent,

- Pls fix the static test
https://travis-ci.org/magento/magento2/jobs/394470047

- Pls fix the Integration Tests
https://travis-ci.org/magento/magento2/jobs/394470051


-  Manual Testing
:white_check_mark: Order and Credimemo Zero Subtotal
:white_check_mark: Order and Credimemo Zero Subtotal mixed

**Comment by larsroettig** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @larsroettig. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @larsroettig, here is your [new Magento instance](https://pr-16005.engcom.dev.magento.com).
Admin access: [https://pr-16005.engcom.dev.magento.com/admin](https://pr-16005.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`

**Comment by viral-wagento** [comment]
@larsroettig 
Thanks for review, I will work on your requested changes and update branch asap.

Currently I am working on Static Test - app/code/Magento/Sales/Model/Order.php:612.
[The method canCreditmemo() has a Cyclomatic Complexity of 10. The configured cyclomatic complexity threshold is 10.]

File: app/code/Magento/Sales/Model/Order.php [hunk 2]
```diff
@@ -614,30 +614,49 @@ public function canCreditmemo()
...
-        if (abs($totalRefunded) < .0001) {
-            return false;
+        if (abs($this->getGrandTotal()) < .0001) {
+            return $this->canCreditmemoforZeroTotal($totalRefunded);
         }
+        $isRefundZero = abs($totalRefunded) < .0001;
         // Case when Adjustment Fee (adjustment_negative) has been used for first creditmemo
-        if (abs($totalRefunded - $this->getAdjustmentNegative()) < .0001) {
+        $hasAdjustmentFee = abs($totalRefunded - $this->getAdjustmentNegative()) < .0001;
+        $hasActinFlag = $this->getActionFlag(self::ACTION_FLAG_EDIT) === false;
```
Comment (sivaschenko) [inline app/code/Magento/Sales/Model/Order.php:634]: Please fix the variable nambe typo: $hasActinFlag -> $hasActionFlag

File: app/code/Magento/Sales/Model/Order.php [hunk 3]
```diff
@@ -614,30 +614,49 @@ public function canCreditmemo()
...
             return false;
         }
+        return true;
+    }
 
-        if ($this->getActionFlag(self::ACTION_FLAG_EDIT) === false) {
+    /**
+     * @param $totalRefunded
+     * @return bool
+     *
```
Comment (sivaschenko) [inline app/code/Magento/Sales/Model/Order.php:644]: Please remove this blank line

File: app/code/Magento/Sales/Model/Order.php [hunk 4]
```diff
@@ -614,30 +614,49 @@ public function canCreditmemo()
...
+    public function canCreditmemoforZeroTotal($totalRefunded)
+    {
+        $totalPaid = $this->getTotalPaid();
+        //check if total paid is less than grandtotal
+        $checkAmtTotalPaid = $totalPaid <= $this->getGrandTotal();
+        //case when amount is due for invoice
+        $dueAmountCondition = $this->canInvoice() && ($checkAmtTotalPaid);
+        //case when paid amount is refunded and order has creditmemo created
+        $paidAmtIsRefunded = $this->getTotalRefunded() == $totalPaid && count($this->getCreditmemosCollection()) > 0;
+        $state1 = $dueAmountCondition || $paidAmtIsRefunded;
```
Comment (sivaschenko) [inline app/code/Magento/Sales/Model/Order.php:655]: Please use meaningful variable names, avoid generic names like `$state1` and `$state2`

File: app/code/Magento/Sales/Model/Order/Creditmemo.php [hunk 5]
```diff
@@ -612,7 +626,18 @@ public function getIncrementId()
...
     public function isValidGrandTotal()
     {
-        return !($this->getGrandTotal() <= 0 && !$this->getAllowZeroGrandTotal());
+        return !($this->getGrandTotal() <= 0 && !$this->isAllowZeroGrandTotal());
+    }
+
+    /**
+     * @return bool
+     */
+
```
Comment (sivaschenko) [inline app/code/Magento/Sales/Model/Order/Creditmemo.php:635]: Please remove the blank line

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @viral-wagento , thanks for your contribution! Please see the review notes, fix failing tests, ensure proper code style in all the changed files (including tests), update the branch and resolve conflicts.

**Comment by viral-wagento** [comment]
Thanks @sivaschenko for reviewing. I will be implementing you changes and update my branch asap.

**Comment by viral-wagento** [comment]
@sivaschenko  @larsroettig and @ishakhsuvarov 

I was trying to work on @sivaschenko 's given changes and conflicts. Then I found that my changes are already there in 2.2-develop branch. 

When did some investigation, I came to know that @SeruyV has copied my code and his code already merged in 2.2-develop now.

@sivaschenko  - you can see that, he has added same variable names, same function name, and has same typo in 'hasActionFlag' variable. He only did some changes after coying my code.

I was working for this, but unfortunately I didn't get credit for this.   ¯ \_(ツ)_/ ¯
He didn't even thought to at least mention a reference of my PR.

Commit from @SeruyV can be seen here: https://github.com/magento/magento2/commit/9fe5066481ba9aee47c921d8e1ac64b18ee7b8c7#diff-6f2b3927e6511c70ebb080c37d84c2d3

**Comment by viral-wagento** [comment]
@SeruyV  - when some juniors developers / new contributors are trying to contribute something, and facing issue, you can simple guide / help them. Magento is an open community to help and grow each other.

I am new to magento 2 and facing cyclomatic complexity issue in this PR, apart from that, my solution was working perfect. 

What you did is, just rearranged code, but the logic and everything is just as mine, even the comments for code (/Magento/Sales/Model/Order.php)  :( :(

**Comment by larsroettig** [comment]
It is already fixed see:
https://github.com/magento/magento2/commit/b613382e6f2baa698e4e6bc9af279a81f698a49e

**Comment by sivaschenko** [comment]
Hi @viral-wagento we are very sorry that your contribution was reused without your credentials. It was first forward ported and merged to 2.3 - https://github.com/magento/magento2/commit/b613382e6f2baa698e4e6bc9af279a81f698a49e and then backported to 2.2. That's why that situation occurred.

I did raise this case with the teams to prevent such situations in the future.

Again, thank you very much for your contribution and sorry for the unfair situation.

**Comment by viral-wagento** [comment]
Thank you for your response, @sivaschenko .!!
Can I get points for this or not?

**Comment by sivaschenko** [comment]
@viral-wagento, unfortunately, we do not have a way to assign points for closed pull requests, but I see your new https://github.com/magento/magento2/pull/19280 that will be rewarded once merged

**Comment by viral-wagento** [comment]
@sivaschenko Okay no problem.
I asked for points of this PR, because this should be considered as new feature.
Anyway, It is fine, if you have no way.
Thanks!

**Pull request closed without merging**